File: passenger-service/src/main/java/io/voitovich/yura/passengerservice/security/service/impl/SecurityServiceImpl.java
package io.voitovich.yura.passengerservice.security.service.impl;

import io.voitovich.yura.passengerservice.exception.PassengerProfileAccessDeniedException;
import io.voitovich.yura.passengerservice.security.service.SecurityService;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.stereotype.Service;

import java.util.UUID;


@Service
public class SecurityServiceImpl implements SecurityService {
    @Override
    public void checkUserAccess(UUID userSub) {
        var token = (Jwt) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        var tokenSub = UUID.fromString(token.getClaim("sub"));
        if (!tokenSub.equals(userSub)) {
            throw new PassengerProfileAccessDeniedException(String
                    .format("Access denied, stored sub id: %s and given sub id: %s are not equal",
                            userSub, tokenSub));
        }

    }
}

File: ride-service/src/main/kotlin/io/voitovich/yura/rideservice/security/SecurityServiceImpl.kt
package io.voitovich.yura.rideservice.security

import io.voitovich.yura.rideservice.exception.PassengerProfileAccessDeniedException
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.stereotype.Service
import java.util.*

@Service
class SecurityServiceImpl : SecurityService {

    companion object {
        const val PASSENGER_PROFILE_ACCESS_DENIED_EXCEPTION_MESSAGE = "Access denied, stored sub id: %s and given sub id: %s are not equal"
    }

    override fun checkUserAccess(userSub: UUID) {
        val token = SecurityContextHolder.getContext().authentication.principal as Jwt
        val tokenSub = UUID.fromString(token.getClaim("sub"))
        if (tokenSub != userSub) {
            throw PassengerProfileAccessDeniedException(
                PASSENGER_PROFILE_ACCESS_DENIED_EXCEPTION_MESSAGE.format(
                    userSub, tokenSub
                )
            )
        }
    }
}
